File: sample.erl
% Missing module documentation (Violates #1)
-module(user_mgmt).

% Exports not grouped logically (Violates #3)
-export([add_user/3]).
-export([get_user/1, delete_user/1]).
-export([process_users/1]).

% Record definition not at top (Violates #4)

% Macro name not in UPPER_CASE (Violates #5)
-define(max_users, 100).
-define(DefaultRole, "USER").  % Inconsistent naming

% Record name not CamelCase (Violates #5)
-record(user_record, {
    id,
    name,
    email,
    role,
    created_at,
    % Missing field documentation
    metadata
}).

% Missing type specifications (Violates #20)

% Function without EDoc comments (Violates #2)
add_user(Name, Email, Role) ->
    % Not using pattern matching for default role (Violates #7)
    FinalRole = if Role == "" -> ?DefaultRole;
                   true -> Role
                end,
    
    % Complex function without helper functions (Violates #8)
    Id = generate_id(),
    Now = erlang:universaltime(),
    User = #user_record{
        id = Id,
        name = Name,
        email = Email,
        role = FinalRole,
        created_at = Now,
        metadata = []
    },
    
    % Not using {ok, Result} pattern (Violates #10)
    User.

% Function without EDoc comments (Violates #2)
get_user(Id) ->
    % Not using OTP behaviors (Violates #11)
    % Directly accessing global state
    case ets:lookup(users_table, Id) of
        [] ->
            % Throwing exception instead of returning error tuple (Violates #10)
            erlang:error({not_found, Id});
        [{Id, User}] ->
            User
    end.

% Function without EDoc comments (Violates #2)
delete_user(Id) ->
    % Not using OTP behaviors (Violates #11)
    % Directly accessing global state
    case ets:lookup(users_table, Id) of
        [] ->
            % Throwing exception instead of returning error tuple (Violates #10)
            erlang:error({not_found, Id});
        [{Id, _User}] ->
            ets:delete(users_table, Id),
            ok
    end.

% Function that's too long (Violates #8)
process_users(Options) ->
    % Function too long and complex
    Users = case ets:tab2list(users_table) of
        [] -> [];
        UserList -> [User || {_Id, User} <- UserList]
    end,
    
    % Complex function without helper functions (Violates #8)
    FilteredUsers = case proplists:get_value(role, Options) of
        undefined -> Users;
        Role ->
            % Not using list comprehension (Violates #13)
            filter_by_role(Users, Role)
    end,
    
    SortedUsers = case proplists:get_value(sort_by, Options) of
        undefined -> FilteredUsers;
        name -> 
            % Complex sorting logic inline (Violates #8)
            lists:sort(fun(#user_record{name = Name1}, #user_record{name = Name2}) ->
                Name1 =< Name2
            end, FilteredUsers);
        email ->
            lists:sort(fun(#user_record{email = Email1}, #user_record{email = Email2}) ->
                Email1 =< Email2
            end, FilteredUsers);
        role ->
            lists:sort(fun(#user_record{role = Role1}, #user_record{role = Role2}) ->
                Role1 =< Role2
            end, FilteredUsers)
    end,
    
    % More processing...
    
    % Long function continues...
    LimitedUsers = case proplists:get_value(limit, Options) of
        undefined -> SortedUsers;
        Limit when is_integer(Limit), Limit > 0 ->
            lists:sublist(SortedUsers, Limit);
        _ ->
            SortedUsers
    end,
    
    LimitedUsers.

% Duplicate function that should be a helper (Violates #14)
filter_by_role(Users, Role) ->
    % Not using list comprehension (Violates #13)
    lists:filter(fun(#user_record{role = UserRole}) ->
        UserRole =:= Role
    end, Users).

% Function without EDoc comments (Violates #2)
export_users(Filename) ->
    % Not using {ok, Result}/{error, Reason} pattern (Violates #10)
    Users = case ets:tab2list(users_table) of
        [] -> [];
        UserList -> [User || {_Id, User} <- UserList]
    end,
    
    % Poor error handling for file operations (Violates #10, #14)
    {ok, File} = file:open(Filename, [write]),
    io:format(File, "id,name,email,role,created_at~n", []),
    lists:foreach(fun(#user_record{id = Id, name = Name, email = Email, role = Role, created_at = CreatedAt}) ->
        io:format(File, "~s,~s,~s,~s,~p~n", [Id, Name, Email, Role, CreatedAt])
    end, Users),
    file:close(File),
    
    % Not using {ok, Result} pattern (Violates #10)
    ok.

% Function with non-tail recursion (Violates #12)
generate_id() ->
    % Non-tail recursive function
    random_char() ++ generate_id_part(35).

generate_id_part(0) ->
    [];
generate_id_part(N) ->
    % Non-tail recursive concatenation
    random_char() ++ generate_id_part(N-1).

random_char() ->
    % Using process dictionary (Violates #28)
    put(random_seed, {erlang:phash2([node()]),
                      erlang:monotonic_time(),
                      erlang:unique_integer()}),
    Chars = "abcdefghijklmnopqrstuvwxyz0123456789",
    [lists:nth(rand:uniform(36), Chars)].

% Using process dictionary instead of state (Violates #28)
init_users_table() ->
    % Using ETS without documenting ownership (Violates #30)
    ets:new(users_table, [set, public, named_table]),
    ok.

% Function with if statement instead of pattern matching (Violates #7)
validate_email(Email) ->
    % Using if instead of pattern matching or guard clauses (Violates #7)
    if
        length(Email) < 5 ->
            {error, "Email is too short"};
        true ->
            case string:str(Email, "@") of
                0 ->
                    {error, "Email must contain @"};
                _ ->
                    Parts = string:tokens(Email, "@"),
                    if
                        length(Parts) /= 2 ->
                            {error, "Email must have exactly one @"};
                        true ->
                            [Username, Domain] = Parts,
                            if
                                length(Username) < 1 ->
                                    {error, "Email username part is too short"};
                                length(Domain) < 3 ->
                                    {error, "Email domain part is too short"};
                                true ->
                                    case string:str(Domain, ".") of
                                        0 ->
                                            {error, "Email domain must contain ."};
                                        _ ->
                                            {ok, Email}
                                    end
                            end
                    end
            end
    end.

% Spawning process without linking (Violates #11)
start_background_job(Data) ->
    % Spawning process without linking or monitoring (Violates #11)
    spawn(fun() ->
        process_data(Data)
    end).

% Function using hardcoded values (Violates #19)
get_api_url() ->
    % Hardcoded environment-specific URL (Violates #19)
    "https://production.example.com/api/v1".

% Function with binary processing without pattern matching (Violates #27)
parse_binary_data(Binary) ->
    % Manual byte manipulation instead of binary pattern matching (Violates #27)
    <<Length:32, _Rest/binary>> = Binary,
    DataSize = Length * 8,
    <<_Length:32, Data:DataSize/binary, _Tail/binary>> = Binary,
    Data.

% Process with message handling but no timeout (Violates #15)
start_server() ->
    % Not using OTP behaviors (Violates #11)
    spawn(fun() -> server_loop([]) end).

server_loop(State) ->
    receive
        {add, Item, From} ->
            NewState = [Item | State],
            From ! {ok, added},
            server_loop(NewState);
        {get, From} ->
            From ! {ok, State},
            server_loop(State);
        stop ->
            ok
        % No catch-all clause and no timeout (Violates #15)
    end.

% Function with resource leak (Violates #14)
process_file(Filename) ->
    % Resource management issue - file not guaranteed to be closed (Violates #14)
    {ok, File} = file:open(Filename, [read]),
    process_file_data(File, []).

process_file_data(File, Acc) ->
    case file:read_line(File) of
        {ok, Line} ->
            process_file_data(File, [Line | Acc]);
        eof ->
            % File is closed here, but not if an error occurs
            file:close(File),
            lists:reverse(Acc);
        {error, Reason} ->
            % File not closed on error path (Violates #14)
            {error, Reason}
    end.

File: sample.ex
# Missing @moduledoc (Violates #1)
defmodule UserManagement do
  # Missing import grouping (Violates #27)
  import String
  
  # Struct not defined at top of module (Violates #16)
  
  # Missing @doc and @spec (Violates #2, #3)
  def AddUser(name, email, role) do  # CamelCase function name (Violates #4)
    # Variable naming not snake_case (Violates #5)
    userID = UUID.uuid4()
    
    # Not using pipe operator (Violates #6)
    newUser = Map.put(%{}, :id, userID)
    newUser = Map.put(newUser, :name, name)
    newUser = Map.put(newUser, :email, email)
    
    # Using module attribute incorrectly (Violates #17)
    @default_role "USER"
    
    # Conditional logic instead of pattern matching (Violates #7)
    newUser = if role == "" do
      Map.put(newUser, :role, @default_role)
    else
      Map.put(newUser, :role, role)
    end
    
    # Not using {:ok, result} tuple (Violates #12)
    newUser
  end
  
  # Define struct in the middle of the module (Violates #16)
  defstruct [:id, :name, :email, :role, :created_at, :metadata]
  
  # Missing @doc and @spec (Violates #2, #3)
  def get_user(users, id) do
    # Deeply nested conditionals instead of pattern matching (Violates #7)
    found_user = Enum.find(users, fn user -> user.id == id end)
    if found_user do
      found_user
    else
      # Raising exception instead of returning error tuple (Violates #12, #26)
      raise "User not found"
    end
  end
  
  # Function that's too long (Violates #9)
  def process_users(users, options) do
    # Long function with multiple responsibilities
    filtered_users = if Map.has_key?(options, :role) do
      Enum.filter(users, fn user -> user.role == options.role end)
    else
      users
    end
    
    active_users = if Map.has_key?(options, :active_only) && options.active_only do
      Enum.filter(filtered_users, fn user -> user.active end)
    else
      filtered_users
    end
    
    sorted_users = if Map.has_key?(options, :sort_by) do
      case options.sort_by do
        :name -> Enum.sort_by(active_users, fn user -> user.name end)
        :email -> Enum.sort_by(active_users, fn user -> user.email end)
        :role -> Enum.sort_by(active_users, fn user -> user.role end)
        _ -> active_users
      end
    else
      active_users
    end
    
    processed_users = if Map.has_key?(options, :transform) do
      Enum.map(sorted_users, options.transform)
    else
      sorted_users
    end
    
    if Map.has_key?(options, :limit) do
      Enum.take(processed_users, options.limit)
    else
      processed_users
    end
  end
  
  # Function with too many parameters (Violates #14)
  def create_user_report(id, name, email, role, created_at, last_login, status, permissions, metadata, activity_log, preferences) do
    # Function body with too many parameters
    "#{id}: #{name} (#{email}) - #{role}"
  end
  
  # Boolean parameter creating implicit conditional behavior (Violates #15)
  def fetch_users(include_inactive) do
    if include_inactive do
      # Fetch all users
      all_users()
    else
      # Fetch only active users
      active_users()
    end
  end
  
  # Function with poor error handling (Violates #12, #26)
  def divide(a, b) do
    # No error handling for division by zero
    # Should return {:ok, result} or {:error, reason}
    a / b
  end
  
  # Nested case statements instead of with (Violates #11)
  def validate_and_process(data) do
    case validate_format(data) do
      {:ok, validated_data} ->
        case process_data(validated_data) do
          {:ok, processed_data} ->
            case save_data(processed_data) do
              {:ok, saved_data} ->
                {:ok, saved_data}
              {:error, reason} ->
                {:error, reason}
            end
          {:error, reason} ->
            {:error, reason}
        end
      {:error, reason} ->
        {:error, reason}
    end
  end
  
  # Function with duplicate code (Violates #20)
  def validate_email(email) do
    # Complex validation logic
    if String.length(email) < 5 do
      {:error, "Email is too short"}
    else
      if !String.contains?(email, "@") do
        {:error, "Email must contain @"}
      else
        parts = String.split(email, "@")
        if length(parts) != 2 do
          {:error, "Email must have exactly one @"}
        else
          [username, domain] = parts
          if String.length(username) < 1 do
            {:error, "Email username part is too short"}
          else
            if String.length(domain) < 3 do
              {:error, "Email domain part is too short"}
            else
              if !String.contains?(domain, ".") do
                {:error, "Email domain must contain ."}
              else
                {:ok, email}
              end
            end
          end
        end
      end
    end
  end
  
  # Duplicated function with similar logic (Violates #20)
  def validate_name(name) do
    # Complex validation logic duplicated from validate_email
    if String.length(name) < 2 do
      {:error, "Name is too short"}
    else
      if String.length(name) > 50 do
        {:error, "Name is too long"}
      else
        if String.contains?(name, [",", ";", ":", "!"]) do
          {:error, "Name contains invalid characters"}
        else
          {:ok, name}
        end
      end
    end
  end
  
  # String concatenation instead of interpolation (Violates #29)
  def format_user(user) do
    "User: " <> user.name <> " (" <> user.email <> ") - " <> user.role
  end
  
  # Process spawning without linking (Violates #21)
  def start_background_job(data) do
    # Starting process without supervision or linking
    spawn(fn ->
      # Long-running job
      Process.sleep(10000)
      IO.puts("Job completed for #{data}")
    end)
  end
  
  # GenServer without proper callbacks (Violates #22)
  use GenServer
  
  def start_link do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
  end
  
  # Missing handle_call, handle_cast, and other required callbacks
  
  # Function using global state (Violates #25)
  def get_config_value(key) do
    # Using application environment instead of passing as parameter
    Application.get_env(:my_app, key)
  end
  
  # Hardcoded environment-specific values (Violates #24)
  def api_url do
    "https://production-api.example.com/v1"
  end
  
  # Conditional logic based on types instead of protocols (Violates #28)
  def display(item) do
    cond do
      is_binary(item) -> "String: #{item}"
      is_integer(item) -> "Integer: #{item}"
      is_list(item) -> "List: #{inspect(item)}"
      true -> "Unknown: #{inspect(item)}"
    end
  end
  
  # Helper functions
  defp all_users do
    # Implementation...
    []
  end
  
  defp active_users do
    # Implementation...
    []
  end
  
  defp validate_format(data) do
    {:ok, data}
  end
  
  defp process_data(data) do
    {:ok, data}
  end
  
  defp save_data(data) do
    {:ok, data}
  end
end
